rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp

```diff
@@ -0,0 +1,105 @@
+  jclass clazz = env->GetObjectClass(self);
+  jfieldID node_name_fid = env->GetFieldID(clazz, "nodeName", "Ljava/lang/String;");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jfieldID node_namespace_fid = env->GetFieldID(clazz, "nodeNamespace", "Ljava/lang/String;");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jfieldID topic_type_fid = env->GetFieldID(clazz, "topicType", "Ljava/lang/String;");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jfieldID endpoint_type_fid = env->GetFieldID(clazz, "endpointType", endpoint_type_enum_path);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jfieldID endpoint_gid_fid = env->GetFieldID(clazz, "endpointGID", "[B");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jfieldID qos_fid = env->GetFieldID(clazz, "qos", "Lorg/ros2/rcljava/qos/QoSProfile;");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+
+  jstring jnode_name = env->NewStringUTF(p->node_name);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  env->SetObjectField(self, node_name_fid, jnode_name);
+  jstring jnode_namespace = env->NewStringUTF(p->node_namespace);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  env->SetObjectField(self, node_namespace_fid, jnode_namespace);
+  jstring jtopic_type = env->NewStringUTF(p->topic_type);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  env->SetObjectField(self, topic_type_fid, jtopic_type);
+  jfieldID enum_value_fid;
+  switch (p->endpoint_type) {
+    case RMW_ENDPOINT_INVALID:
+      enum_value_fid = env->GetStaticFieldID(
+        endpoint_type_clazz, "INVALID", endpoint_type_enum_path);
+      break;
+    case RMW_ENDPOINT_PUBLISHER:
+      enum_value_fid = env->GetStaticFieldID(
+        endpoint_type_clazz, "PUBLISHER", endpoint_type_enum_path);
+      break;
+    case RMW_ENDPOINT_SUBSCRIPTION:
+      enum_value_fid = env->GetStaticFieldID(
+        endpoint_type_clazz, "SUBSCRIPTION", endpoint_type_enum_path);
+      break;
+    default:
+      rcljava_throw_exception(
+        env, "java/lang/IllegalArgumentException", "unknown endpoint type");
+      break;
+  }
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jobject enum_value = env->GetStaticObjectField(endpoint_type_clazz, enum_value_fid);
+  env->SetObjectField(self, endpoint_type_fid, enum_value);
+  jbyteArray jgid = env->NewByteArray(RMW_GID_STORAGE_SIZE);
+  if (jgid == NULL) {
+    rcljava_throw_exception(
+      env, "java/lang/OutOfMemoryError", "cannot allocate java gid byte array");
+    return;
+  }
+  jbyte * gid_content = env->GetByteArrayElements(jgid, nullptr);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  for (size_t i = 0; i < RMW_GID_STORAGE_SIZE; ++i) {
+    gid_content[i] = p->endpoint_gid[i];
+  }
+  env->ReleaseByteArrayElements(jgid, gid_content, 0);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  env->SetObjectField(self, endpoint_gid_fid, jgid);
+  jclass qos_clazz = env->FindClass("org/ros2/rcljava/qos/QoSProfile");
+  jmethodID qos_init_mid = env->GetMethodID(qos_clazz, "<init>", "()V");
```

Comment:
Should we check for exceptions after each of these lines?

Reply:
Addressed in https://github.com/osrf/ros2_java/pull/26/commits/0024d68a62ed3809ff5cbaaccce06467f4e00f04.